File: platform/polyglot/cache_json_awss3_storage/py/cache_json_awss3_storage/cached_http_client.py
"""
Cached HTTP Client — Higher-Order Function Pattern

Wraps any supported HTTP client with S3-backed response caching.
Provides standard HTTP methods where each response has an explicit
``.save_response()`` method for manual cache persistence.

Usage::

    cached = await with_s3_cache(http_client)(s3_config)
    res = await cached.get("https://api.example.com/data")
    key = await res.save_response()   # persist to S3

Supported clients:
    - fetch_client/FetchClient  (response["data"], response["status"])
    - fetch_httpx/AsyncClient   (response.json(), response.status_code)
    - raw httpx.AsyncClient     (response.json(), response.status_code)
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from cache_json_awss3_storage.client_factory import ClientAsync, get_client_factory
from cache_json_awss3_storage.exceptions import JsonS3StorageConfigError
from cache_json_awss3_storage.key_generator import generate_key_string
from cache_json_awss3_storage.logger import create as create_logger
from cache_json_awss3_storage.logger import LoggerProtocol
from cache_json_awss3_storage.storage import JsonS3Storage, create_storage


# ---------------------------------------------------------------------------
# S3CacheConfig
# ---------------------------------------------------------------------------

@dataclass
class S3CacheConfig:
    """Configuration for S3 cache connection."""

    bucket_name: str
    region_name: str | None = None
    endpoint_url: str | None = None
    aws_access_key_id: str | None = None
    aws_secret_access_key: str | None = None
    ttl: float = 600.0
    prefix: str = "cache:"
    addressing_style: str = "path"


# ---------------------------------------------------------------------------
# NormalizedResponse (internal)
# ---------------------------------------------------------------------------

@dataclass
class _NormalizedResponse:
    """Internal unified response shape from adapters."""

    status: int
    headers: dict[str, str]
    body: Any
    ok: bool
    url: str
    method: str


# ---------------------------------------------------------------------------
# HttpClientAdapter protocol (internal)
# ---------------------------------------------------------------------------

@runtime_checkable
class _HttpClientAdapter(Protocol):
    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        body: Any | None = None,
        params: dict[str, str] | None = None,
        timeout: float | None = None,
    ) -> _NormalizedResponse: ...


# ---------------------------------------------------------------------------
# FetchClientAdapter — wraps polyglot fetch_client/FetchClient
# ---------------------------------------------------------------------------

class _FetchClientAdapter:
    """Adapter for polyglot fetch_client.FetchClient."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        body: Any | None = None,
        params: dict[str, str] | None = None,
        timeout: float | None = None,
    ) -> _NormalizedResponse:
        m = method.lower()
        kwargs: dict[str, Any] = {}
        if headers:
            kwargs["headers"] = headers
        if timeout:
            kwargs["timeout"] = timeout

        if m == "get":
            response = await self._client.get(url, params, **kwargs)
        elif m == "delete":
            response = await self._client.delete(url, **kwargs)
        else:
            # post, put, patch — body as second positional arg
            fn = getattr(self._client, m)
            response = await fn(url, body, **kwargs)

        return _NormalizedResponse(
            status=response["status"],
            headers=response.get("headers", {}),
            body=response["data"],
            ok=response.get("ok", 200 <= response["status"] < 300),
            url=url,
            method=method.upper(),
        )


# ---------------------------------------------------------------------------
# FetchHttpxAdapter — wraps fetch_httpx/AsyncClient
# ---------------------------------------------------------------------------

class _FetchHttpxAdapter:
    """Adapter for fetch_httpx.AsyncClient (and raw httpx.AsyncClient)."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        body: Any | None = None,
        params: dict[str, str] | None = None,
        timeout: float | None = None,
    ) -> _NormalizedResponse:
        kwargs: dict[str, Any] = {}
        if headers:
            kwargs["headers"] = headers
        if params:
            kwargs["params"] = params
        if timeout:
            kwargs["timeout"] = timeout
        if body is not None:
            kwargs["json"] = body

        response = await self._client.request(method.upper(), url, **kwargs)

        # Parse body — httpx Response has .json() (sync) and .text (property)
        resp_body: Any = None
        try:
            resp_body = response.json()
        except Exception:
            try:
                resp_body = response.text
            except Exception:
                resp_body = None

        # Convert headers to plain dict
        resp_headers: dict[str, str] = {}
        if hasattr(response.headers, "items"):
            resp_headers = dict(response.headers.items())

        status_code: int = response.status_code
        ok = 200 <= status_code < 300

        return _NormalizedResponse(
            status=status_code,
            headers=resp_headers,
            body=resp_body,
            ok=ok,
            url=url,
            method=method.upper(),
        )


# ---------------------------------------------------------------------------
# detect_adapter — auto-detect HTTP client type
# ---------------------------------------------------------------------------

def _detect_adapter(http_client: Any) -> _HttpClientAdapter:
    """Auto-detect the HTTP client type and return the appropriate adapter."""
    if http_client is None:
        raise JsonS3StorageConfigError(
            "http_client must be a FetchClient, fetch_httpx.AsyncClient, or httpx.AsyncClient"
        )

    # fetch_httpx / raw httpx.AsyncClient — has aclose(), request() takes (method, url)
    if hasattr(http_client, "aclose") and hasattr(http_client, "request"):
        return _FetchHttpxAdapter(http_client)

    # polyglot FetchClient — has .get/.post/.request/.close, dict-style responses
    if (
        hasattr(http_client, "get")
        and hasattr(http_client, "post")
        and hasattr(http_client, "request")
        and hasattr(http_client, "close")
        and not hasattr(http_client, "aclose")
    ):
        return _FetchClientAdapter(http_client)

    raise JsonS3StorageConfigError(
        "Unrecognized HTTP client type. "
        "Supported: FetchClient, fetch_httpx.AsyncClient, httpx.AsyncClient"
    )


# ---------------------------------------------------------------------------
# CachedResponse
# ---------------------------------------------------------------------------

class CachedResponse:
    """
    Wraps a normalized HTTP response with S3 cache persistence.

    Attributes:
        status: HTTP status code.
        headers: Response headers as a plain dict.
        data: Parsed response body (JSON or raw).
        ok: True if status is 2xx.
        url: Request URL.
        method: HTTP method used.
    """

    __slots__ = (
        "status", "headers", "data", "ok", "url", "method",
        "_storage", "_default_ttl", "_logger",
    )

    def __init__(
        self,
        normalized: _NormalizedResponse,
        storage: JsonS3Storage,
        logger: LoggerProtocol,
        default_ttl: float | None = None,
    ) -> None:
        self.status: int = normalized.status
        self.headers: dict[str, str] = normalized.headers
        self.data: Any = normalized.body
        self.ok: bool = normalized.ok
        self.url: str = normalized.url
        self.method: str = normalized.method
        self._storage = storage
        self._default_ttl = default_ttl
        self._logger = logger

    async def save_response(
        self,
        key: str | None = None,
        *,
        ttl: float | None = None,
    ) -> str:
        """
        Manually save this response to S3.

        Args:
            key: Custom cache key. Defaults to hash of ``{method}:{url}``.
            ttl: TTL override in seconds. Defaults to config TTL.

        Returns:
            The S3 storage key used.
        """
        cache_key = key or generate_key_string(f"{self.method}:{self.url}")
        effective_ttl = ttl if ttl is not None else self._default_ttl

        start_time = time.time()

        self._logger.debug(
            f"save_response: method={self.method} url={self.url} key={cache_key}"
        )

        payload: dict[str, Any] = {
            "status": self.status,
            "headers": self.headers,
            "data": self.data,
            "ok": self.ok,
            "url": self.url,
            "method": self.method,
            "cached_at": time.time(),
        }

        saved_key = await self._storage.save(
            cache_key, payload, ttl=effective_ttl
        )

        elapsed = (time.time() - start_time) * 1000
        self._logger.info(
            f"save_response: saved key={saved_key} in {elapsed:.1f}ms"
        )

        return saved_key


# ---------------------------------------------------------------------------
# CachedHttpClient
# ---------------------------------------------------------------------------

class CachedHttpClient:
    """
    HTTP client wrapping an adapter with S3 cache storage.

    Provides standard async HTTP methods (get, post, put, patch, delete).
    Each response is a ``CachedResponse`` with a ``.save_response()`` method.
    """

    def __init__(
        self,
        adapter: _HttpClientAdapter,
        storage: JsonS3Storage,
        client_ctx: Any,
        s3_client: Any,
        logger: LoggerProtocol,
        ttl: float | None = None,
    ) -> None:
        self._adapter = adapter
        self._storage = storage
        self._client_ctx = client_ctx
        self._s3_client = s3_client
        self._logger = logger
        self._ttl = ttl

    async def _request(
        self,
        method: str,
        url: str,
        *,
        headers: dict[str, str] | None = None,
        body: Any | None = None,
        params: dict[str, str] | None = None,
        timeout: float | None = None,
    ) -> CachedResponse:
        start_time = time.time()
        self._logger.debug(f"request: {method} {url}")

        normalized = await self._adapter.request(
            method, url,
            headers=headers, body=body, params=params, timeout=timeout,
        )

        elapsed = (time.time() - start_time) * 1000
        self._logger.info(
            f"request: {method} {url} → {normalized.status} in {elapsed:.1f}ms"
        )

        return CachedResponse(normalized, self._storage, self._logger, self._ttl)

    async def get(
        self,
        url: str,
        params: dict[str, str] | None = None,
        options: dict[str, Any] | None = None,
    ) -> CachedResponse:
        opts = options or {}
        return await self._request(
            "GET", url, params=params,
            headers=opts.get("headers"), timeout=opts.get("timeout"),
        )

    async def post(
        self,
        url: str,
        body: Any | None = None,
        options: dict[str, Any] | None = None,
    ) -> CachedResponse:
        opts = options or {}
        return await self._request(
            "POST", url, body=body,
            headers=opts.get("headers"), timeout=opts.get("timeout"),
        )

    async def put(
        self,
        url: str,
        body: Any | None = None,
        options: dict[str, Any] | None = None,
    ) -> CachedResponse:
        opts = options or {}
        return await self._request(
            "PUT", url, body=body,
            headers=opts.get("headers"), timeout=opts.get("timeout"),
        )

    async def patch(
        self,
        url: str,
        body: Any | None = None,
        options: dict[str, Any] | None = None,
    ) -> CachedResponse:
        opts = options or {}
        return await self._request(
            "PATCH", url, body=body,
            headers=opts.get("headers"), timeout=opts.get("timeout"),
        )

    async def delete(
        self,
        url: str,
        options: dict[str, Any] | None = None,
    ) -> CachedResponse:
        opts = options or {}
        return await self._request(
            "DELETE", url,
            headers=opts.get("headers"), timeout=opts.get("timeout"),
        )

    async def destroy(self) -> None:
        """Cleanup S3 client resources."""
        self._logger.info("destroy: closing S3 client and storage")
        await self._storage.close()
        if self._client_ctx:
            await self._client_ctx.__aexit__(None, None, None)


# ---------------------------------------------------------------------------
# with_s3_cache — the HOF
# ---------------------------------------------------------------------------

def with_s3_cache(
    http_client: Any,
):
    """
    Higher-order function that wraps an HTTP client with S3 caching.

    First call: accepts any supported HTTP client, returns an async config function.
    Second call (async): accepts S3CacheConfig, returns a ready-to-use CachedHttpClient.

    Example::

        import httpx
        from cache_json_awss3_storage import with_s3_cache, S3CacheConfig

        async_client = httpx.AsyncClient()
        cached = await with_s3_cache(async_client)(S3CacheConfig(
            bucket_name="my-cache",
            endpoint_url="http://localhost:4566",
            ttl=600,
        ))

        res = await cached.get("https://api.example.com/users")
        print(res.data)
        await res.save_response()    # cache to S3
        await cached.destroy()       # cleanup

    Args:
        http_client: A FetchClient, fetch_httpx.AsyncClient, or httpx.AsyncClient.

    Returns:
        An async callable that accepts S3CacheConfig and returns CachedHttpClient.
    """
    adapter = _detect_adapter(http_client)

    async def _configure(config: S3CacheConfig) -> CachedHttpClient:
        if not config.bucket_name:
            raise JsonS3StorageConfigError("S3CacheConfig.bucket_name is required")

        logger = create_logger("cache_json_awss3_storage", __file__)
        logger.info(
            f"with_s3_cache: creating cached client "
            f"bucket={config.bucket_name}, region={config.region_name}, "
            f"endpoint={config.endpoint_url}, ttl={config.ttl}"
        )

        # Create S3 client config via existing factory
        client_config = get_client_factory(
            bucket_name=config.bucket_name,
            region_name=config.region_name,
            endpoint_url=config.endpoint_url,
            aws_access_key_id=config.aws_access_key_id,
            aws_secret_access_key=config.aws_secret_access_key,
            addressing_style=config.addressing_style,
            ttl=config.ttl,
        )

        # Open async S3 client via context manager
        client_ctx = ClientAsync(client_config)
        s3_client = await client_ctx.__aenter__()

        # Create JsonS3Storage for cache persistence
        storage = create_storage(
            s3_client,
            config.bucket_name,
            key_prefix=config.prefix,
            ttl=config.ttl,
            region=config.region_name,
            logger=logger,
        )

        return CachedHttpClient(
            adapter=adapter,
            storage=storage,
            client_ctx=client_ctx,
            s3_client=s3_client,
            logger=logger,
            ttl=config.ttl,
        )

    return _configure

File: platform/polyglot/cache_json_awss3_storage/mjs/src/cached-http-client.ts
/**
 * Cached HTTP Client — Higher-Order Function Pattern
 *
 * Wraps any supported HTTP client with S3-backed response caching.
 * Provides standard HTTP methods where each response has an explicit
 * `.saveResponse()` method for manual cache persistence.
 *
 * Usage:
 *   const cached = withS3Cache(httpClient)(s3Config);
 *   const res = await cached.get("https://api.example.com/data");
 *   const key = await res.saveResponse();  // persist to S3
 *
 * Supported clients:
 *   - fetch_client/FetchClient  (response.data, response.status)
 *   - fetch_undici/AsyncClient  (response.json(), response.statusCode)
 *   - raw undici Dispatcher     (wrapped in AsyncClient internally)
 */

import { getClientFactory, createAsyncClient } from "./client-factory.js";
import type { AsyncClientHandle } from "./client-factory.js";
import { JsonS3Storage, createStorage } from "./storage.js";
import { generateKeyString } from "./key-generator.js";
import {
  JsonS3StorageError,
  JsonS3StorageConfigError,
} from "./errors.js";
import { create as createLogger, NullLogger } from "./logger.js";
import type { Logger } from "./logger.js";

// ---------------------------------------------------------------------------
// S3CacheConfig
// ---------------------------------------------------------------------------

export interface S3CacheConfig {
  bucketName: string;
  regionName?: string;
  endpointUrl?: string;
  awsAccessKeyId?: string;
  awsSecretAccessKey?: string;
  ttl?: number;
  prefix?: string;
  addressingStyle?: "path" | "virtual";
}

// ---------------------------------------------------------------------------
// RequestOptions
// ---------------------------------------------------------------------------

export interface RequestOptions {
  headers?: Record<string, string>;
  timeout?: number;
}

// ---------------------------------------------------------------------------
// NormalizedResponse (internal)
// ---------------------------------------------------------------------------

interface NormalizedResponse {
  status: number;
  headers: Record<string, string>;
  body: unknown;
  ok: boolean;
  url: string;
  method: string;
}

// ---------------------------------------------------------------------------
// HttpClientAdapter (internal)
// ---------------------------------------------------------------------------

interface HttpClientAdapter {
  request(
    method: string,
    url: string,
    options?: {
      headers?: Record<string, string>;
      body?: unknown;
      params?: Record<string, string>;
      timeout?: number;
    }
  ): Promise<NormalizedResponse>;
}

// ---------------------------------------------------------------------------
// FetchClientAdapter — wraps polyglot fetch_client/FetchClient
// ---------------------------------------------------------------------------

class FetchClientAdapter implements HttpClientAdapter {
  constructor(private readonly client: any) {}

  async request(
    method: string,
    url: string,
    options?: {
      headers?: Record<string, string>;
      body?: unknown;
      params?: Record<string, string>;
      timeout?: number;
    }
  ): Promise<NormalizedResponse> {
    const m = method.toLowerCase();
    let response: any;

    const reqOpts: Record<string, unknown> = {};
    if (options?.headers) reqOpts.headers = options.headers;
    if (options?.timeout) reqOpts.timeout = options.timeout;

    if (m === "get") {
      response = await this.client.get(url, options?.params, reqOpts);
    } else if (m === "delete") {
      response = await this.client.delete(url, reqOpts);
    } else {
      // post, put, patch — body as second arg
      response = await this.client[m](url, options?.body, reqOpts);
    }

    return {
      status: response.status,
      headers: response.headers ?? {},
      body: response.data,
      ok: response.ok,
      url,
      method: method.toUpperCase(),
    };
  }
}

// ---------------------------------------------------------------------------
// UndiciClientAdapter — wraps fetch_undici/AsyncClient
// ---------------------------------------------------------------------------

class UndiciClientAdapter implements HttpClientAdapter {
  constructor(private readonly client: any) {}

  async request(
    method: string,
    url: string,
    options?: {
      headers?: Record<string, string>;
      body?: unknown;
      params?: Record<string, string>;
      timeout?: number;
    }
  ): Promise<NormalizedResponse> {
    const reqOpts: Record<string, unknown> = {};
    if (options?.headers) reqOpts.headers = options.headers;
    if (options?.timeout) reqOpts.timeout = options.timeout;
    if (options?.body !== undefined) reqOpts.json = options.body;
    if (options?.params) reqOpts.params = options.params;

    const response = await this.client.request(method.toUpperCase(), url, reqOpts);

    // fetch_undici Response has lazy body — consume via .json()
    let body: unknown;
    try {
      body = await response.json();
    } catch {
      try {
        body = await response.text();
      } catch {
        body = null;
      }
    }

    // Convert Headers object to plain Record
    const headers: Record<string, string> = {};
    if (response.headers) {
      if (typeof response.headers.entries === "function") {
        for (const [k, v] of response.headers.entries()) {
          headers[k] = v;
        }
      } else if (typeof response.headers.forEach === "function") {
        response.headers.forEach((v: string, k: string) => {
          headers[k] = v;
        });
      }
    }

    return {
      status: response.statusCode,
      headers,
      body,
      ok: response.ok ?? (response.statusCode >= 200 && response.statusCode < 300),
      url,
      method: method.toUpperCase(),
    };
  }
}

// ---------------------------------------------------------------------------
// UndiciDispatcherAdapter — wraps raw undici Dispatcher
// ---------------------------------------------------------------------------

class UndiciDispatcherAdapter implements HttpClientAdapter {
  private innerAdapter: UndiciClientAdapter | null = null;

  constructor(private readonly dispatcher: any) {}

  async request(
    method: string,
    url: string,
    options?: {
      headers?: Record<string, string>;
      body?: unknown;
      params?: Record<string, string>;
      timeout?: number;
    }
  ): Promise<NormalizedResponse> {
    // Lazy-init: wrap Dispatcher in an AsyncClient from fetch_undici
    if (!this.innerAdapter) {
      // Dynamic import to avoid hard dependency
      try {
        // Dynamic import — fetch_undici is optional, not a direct dependency
        const mod = await (import("fetch_undici" as string) as Promise<any>);
        const asyncClient = new mod.AsyncClient({ dispatcher: this.dispatcher });
        this.innerAdapter = new UndiciClientAdapter(asyncClient);
      } catch {
        throw new JsonS3StorageConfigError(
          "Cannot create AsyncClient from Dispatcher — fetch_undici package not available"
        );
      }
    }
    return this.innerAdapter.request(method, url, options);
  }
}

// ---------------------------------------------------------------------------
// detectAdapter — auto-detect HTTP client type
// ---------------------------------------------------------------------------

function detectAdapter(httpClient: unknown): HttpClientAdapter {
  if (!httpClient || typeof httpClient !== "object") {
    throw new JsonS3StorageConfigError(
      "httpClient must be a FetchClient, AsyncClient, or undici Dispatcher"
    );
  }

  const c = httpClient as any;

  // fetch_undici/AsyncClient — has circuitBreaker property + request method
  if ("circuitBreaker" in c && typeof c.request === "function") {
    return new UndiciClientAdapter(c);
  }

  // polyglot FetchClient — has .get/.post/.request/.stream/.close, no circuitBreaker
  if (
    typeof c.get === "function" &&
    typeof c.post === "function" &&
    typeof c.request === "function" &&
    typeof c.close === "function" &&
    !("circuitBreaker" in c)
  ) {
    return new FetchClientAdapter(c);
  }

  // Raw undici Dispatcher — has .dispatch method
  if (typeof c.dispatch === "function") {
    return new UndiciDispatcherAdapter(c);
  }

  throw new JsonS3StorageConfigError(
    "Unrecognized HTTP client type. Supported: FetchClient, AsyncClient, undici Dispatcher"
  );
}

// ---------------------------------------------------------------------------
// CachedResponse
// ---------------------------------------------------------------------------

export class CachedResponse {
  readonly status: number;
  readonly headers: Record<string, string>;
  readonly data: unknown;
  readonly ok: boolean;
  readonly url: string;
  readonly method: string;

  private readonly _storage: JsonS3Storage;
  private readonly _defaultTtl?: number;
  private readonly _logger: Logger;

  constructor(
    normalized: NormalizedResponse,
    storage: JsonS3Storage,
    logger: Logger,
    defaultTtl?: number,
  ) {
    this.status = normalized.status;
    this.headers = normalized.headers;
    this.data = normalized.body;
    this.ok = normalized.ok;
    this.url = normalized.url;
    this.method = normalized.method;
    this._storage = storage;
    this._defaultTtl = defaultTtl;
    this._logger = logger;
  }

  /**
   * Manually save this response to S3.
   *
   * @param key - Custom cache key. Defaults to hash of `${method}:${url}`.
   * @param options - Optional overrides (ttl).
   * @returns The S3 storage key used.
   */
  async saveResponse(
    key?: string,
    options?: { ttl?: number },
  ): Promise<string> {
    const cacheKey = key ?? generateKeyString(`${this.method}:${this.url}`);
    const ttl = options?.ttl ?? this._defaultTtl;

    const startTime = performance.now();

    this._logger.debug(
      `saveResponse: method=${this.method} url=${this.url} key=${cacheKey}`
    );

    const payload: Record<string, unknown> = {
      status: this.status,
      headers: this.headers,
      data: this.data,
      ok: this.ok,
      url: this.url,
      method: this.method,
      cached_at: Date.now() / 1000,
    };

    const savedKey = await this._storage.save(cacheKey, payload, { ttl });

    const elapsed = performance.now() - startTime;
    this._logger.info(
      `saveResponse: saved key=${savedKey} in ${elapsed.toFixed(1)}ms`
    );

    return savedKey;
  }
}

// ---------------------------------------------------------------------------
// CachedHttpClient
// ---------------------------------------------------------------------------

export class CachedHttpClient {
  private readonly _adapter: HttpClientAdapter;
  private readonly _storage: JsonS3Storage;
  private readonly _clientHandle: AsyncClientHandle;
  private readonly _logger: Logger;
  private readonly _ttl?: number;

  constructor(
    adapter: HttpClientAdapter,
    storage: JsonS3Storage,
    clientHandle: AsyncClientHandle,
    logger: Logger,
    ttl?: number,
  ) {
    this._adapter = adapter;
    this._storage = storage;
    this._clientHandle = clientHandle;
    this._logger = logger;
    this._ttl = ttl;
  }

  private async _request(
    method: string,
    url: string,
    options?: {
      headers?: Record<string, string>;
      body?: unknown;
      params?: Record<string, string>;
      timeout?: number;
    },
  ): Promise<CachedResponse> {
    const startTime = performance.now();
    this._logger.debug(`request: ${method} ${url}`);

    const normalized = await this._adapter.request(method, url, options);

    const elapsed = performance.now() - startTime;
    this._logger.info(
      `request: ${method} ${url} → ${normalized.status} in ${elapsed.toFixed(1)}ms`
    );

    return new CachedResponse(normalized, this._storage, this._logger, this._ttl);
  }

  async get(
    url: string,
    params?: Record<string, string>,
    options?: RequestOptions,
  ): Promise<CachedResponse> {
    return this._request("GET", url, { ...options, params });
  }

  async post(
    url: string,
    body?: unknown,
    options?: RequestOptions,
  ): Promise<CachedResponse> {
    return this._request("POST", url, { ...options, body });
  }

  async put(
    url: string,
    body?: unknown,
    options?: RequestOptions,
  ): Promise<CachedResponse> {
    return this._request("PUT", url, { ...options, body });
  }

  async patch(
    url: string,
    body?: unknown,
    options?: RequestOptions,
  ): Promise<CachedResponse> {
    return this._request("PATCH", url, { ...options, body });
  }

  async delete(
    url: string,
    options?: RequestOptions,
  ): Promise<CachedResponse> {
    return this._request("DELETE", url, options);
  }

  /**
   * Cleanup S3 client resources. Call when done using the cached client.
   */
  async destroy(): Promise<void> {
    this._logger.info("destroy: closing S3 client and storage");
    await this._storage.close();
    this._clientHandle.destroy();
  }
}

// ---------------------------------------------------------------------------
// withS3Cache — the HOF
// ---------------------------------------------------------------------------

/**
 * Higher-order function that wraps an HTTP client with S3 caching.
 *
 * First call: accepts any supported HTTP client, returns a config function.
 * Second call: accepts S3CacheConfig, returns a ready-to-use CachedHttpClient.
 *
 * @example
 * ```typescript
 * import { withS3Cache } from "cache_json_awss3_storage";
 * import { FetchClient } from "fetch_client";
 *
 * const client = FetchClient.create(config);
 * const cached = withS3Cache(client)({
 *   bucketName: "my-cache",
 *   endpointUrl: "http://localhost:4566",
 *   ttl: 600,
 * });
 *
 * const res = await cached.get("https://api.example.com/users");
 * console.log(res.data);
 * await res.saveResponse();     // cache to S3
 * await cached.destroy();       // cleanup
 * ```
 */
export function withS3Cache(
  httpClient: unknown,
): (config: S3CacheConfig) => CachedHttpClient {
  const adapter = detectAdapter(httpClient);

  return (config: S3CacheConfig): CachedHttpClient => {
    if (!config.bucketName) {
      throw new JsonS3StorageConfigError("S3CacheConfig.bucketName is required");
    }

    const logger = createLogger("cache_json_awss3_storage", import.meta.url);
    logger.info(
      `withS3Cache: creating cached client bucket=${config.bucketName}, ` +
        `region=${config.regionName}, endpoint=${config.endpointUrl}, ttl=${config.ttl}`
    );

    // Create S3 client via existing factory
    const clientConfig = getClientFactory({
      bucketName: config.bucketName,
      regionName: config.regionName,
      endpointUrl: config.endpointUrl,
      awsAccessKeyId: config.awsAccessKeyId,
      awsSecretAccessKey: config.awsSecretAccessKey,
      addressingStyle: config.addressingStyle ?? "path",
      ttl: config.ttl ?? 600,
    });

    const clientHandle = createAsyncClient(clientConfig);

    // Create JsonS3Storage for cache persistence
    const storage = createStorage({
      s3Client: clientHandle.client,
      bucketName: config.bucketName,
      keyPrefix: config.prefix ?? "cache:",
      ttl: config.ttl ?? 600,
      region: config.regionName,
      logger,
    });

    return new CachedHttpClient(adapter, storage, clientHandle, logger, config.ttl);
  };
}
